util/unifyfs-stage/src/unifyfs-stage-transfer.c

```diff
@@ -0,0 +1,403 @@
+{
+    int ret = 0;
+    int count = 0;
+    FILE* fp = NULL;
+    char* src = NULL;
+    char* dst = NULL;
+    char linebuf[LINE_MAX] = { 0, };
+    struct stat sb = { 0, };
+
+    if (!ctx) {
+        return EINVAL;
+    }
+
+    fp = fopen(ctx->manifest_file, "r");
+    if (!fp) {
+        fprintf(stderr, "failed to open file %s: %s\n",
+		ctx->manifest_file, strerror(errno));
+        ret = errno;
+        goto out;
+    }
+
+    while (NULL != fgets(linebuf, LINE_MAX-1, fp)) {
+        if (strlen(linebuf) < 5) {
+	    // the manifest file perhaps ends with a couple of characters
+	    // and/or a newline not meant to be a transfer spec.
+            if (linebuf[0] == '\n') {
+	        goto out;
+            } else{
+	        fprintf(stderr, "Short (bad) manifest file line: >%s<\n",
+			linebuf);
+		ret = -EINVAL;
+                goto out;
+            }
+	}
+	ret = unifyfs_parse_manifest_line(linebuf, &src, &dst);
+	if (ret < 0) {
+	    fprintf(stderr, "failed to parse %s (%s)\n",
+		    linebuf, strerror(ret));
+            goto out;
+	}
+        if (ctx->mode == UNIFYFS_STAGE_SERIAL) {
+            if (count % total_ranks == rank) {
+                if (verbose) {
+                    fprintf(stdout, "[%d] serial transfer: src=%s, dst=%s\n",
+                                    rank, src, dst);
+                }
+
+                ret = unifyfs_transfer_file_serial(src, dst);
+                if (ret) {
+                    goto out;
+                }
+
+                if (ret < 0) {
+                    fprintf(stderr, "stat on %s failed (err=%d, %s)\n",
+                                    dst, errno, strerror(errno));
+                    ret = errno;
+                    goto out;
+                }
+
+                if (ctx->checksum) {
+                    ret = verify_checksum(src, dst);
+                    if (ret) {
+                        fprintf(stderr, "checksums for >%s< and >%s< differ!\n",
+                                src, dst);
+                        goto out;
+                    }
+                }
+            }
+        } else {
+            if (0 == rank) {
+                int fd = -1;
+
+                if (verbose) {
+                    fprintf(stdout, "[%d] parallel transfer: src=%s, dst=%s\n",
+                                    rank, src, dst);
+                }
+
+                /* FIXME: Since we cannot use the mpi barrier inside the
+                 * unifyfs_transfer_file_parallel(), we need to create the file
+                 * here before others start to access. It would be better if we
+                 * can skip this step.
+                 */
+                fd = open(dst, O_WRONLY|O_CREAT|O_TRUNC, 0600);
+                if (fd < 0) {
+                    fprintf(stderr, "[%d] failed to create the file %s\n",
+                                    rank, dst);
+                    goto out;
+                }
+
+                close(fd);
```

Comment:
I'm not clear on why this empty file needs to be created

Reply:
@tonyhutter Otherwise, all ranks will try to open the target file with O_CREAT inside the unifyfs_trasnfer function. This leads to an incomplete/corrupted file in PFS.

Reply:
I assume this is resolved then?

Reply:
I guess so, because we cannot synchronize mpi processes within the transfer api function, unless I missed something, or there's other simple fix. @tonyhutter

Reply:
Did you try having `do_transfer_file_parallel()` `open(O_CREAT|O_EXCL...)`, and if that doesn't work, `open(O_WRONLY)`?

Reply:
@tonyhutter I think that will work, so something like:

```
do _transfer_file_parallel()
{
    fd = open(O_CREAT|O_EXCL ...); // possibly only single process will succeed

    if (fd < 0 && errno = EEXIST) { // for others
        fd = open(O_WRONLY );
    }
}
```

although we would still need one of the barriers afterwards.

Reply:
Do we need to do anything with this before the PR is ready to upload?  

If the only thing we're saving is not creating an empty (lock?) file, then that doesn't seem like it's worth the effort to stop and do the same thing with a different method.

Reply:
@sandrain yep, that's what I was thinking.  @craigsteffen given that the PR adds a comment saying "FIXME:  ...it would be better if we could fix this step", I think we should fix it in this PR.  Then we could delete that comment.

Reply:
Ok, I can change this and re-push it. 

As long as I'm doing that, anything else you'd like me to work on Tony, or can the rest be pushed the way it is now?

Reply:
Ok, I've read this discussion more carefully this time, but I'm confused.  
The comment says "Since we cannot use the mpi barrier inside the unifyfs_transfer_file_parallel()".  Why can't we?  Why can't we just MPI_File_Open inside the unifyfs_transfer_file_paralle() using the GLOBAL communicator?  We know we're in an MPI context.  That's the natural thing to do.  
If for some reason we can't do that, then I propose we leave the code like it is and remove the FIXME comment.  I don't see any particular reason not to code it that way.  One rank will make a metadata call to create the file, then lots of ranks will open it, and they're all opening it expecting it to be there so they aren't all hitting errors and then having to open it up again.

Reply:
@craigsteffen that's fine.  The main reason I was suggesting using O_EXCL was because the comment made it sound like the "rank 0 creates the file first" way was undesirable, but if you guys are ok with it then leave it as it.

Reply:
comment removed in upcoming patch.